feat(update-user-password):user updating password

Ruthuwamahoro opened the pull request:
-users can update their password when they have validly registered -users must provide their old password, new password and confirm their password

issue in referencing?

Questions:

What does this PR do?

Password Change Functionality Implementation

Description of Task to be completed?
- Ensuring that the user attempting to change the password is a valid registered user with a validated account
- Requires the user to provide their current password for authentication purposes, along with the new password they want to set  and confirmation to avoid errors due to mist, typed password

-For security reasons, ensure the new password is securely hashed before updating it in the database.
- Prevents unauthorised password changes and ensures that the user can only change their password if they correctly provide their old password and confirm the new password.

#### How should this be manually tested?

-run "npm run install"
-run "npm run migrate"
-run "npm run seed"
-run "npm run dev"
 

 background context you want to provide?

Not applicable

#### What are the relevant pivotal tracker/Trello stories?

- id of the project
  #187300151

#### Questions:

-Not applicable

fab-ryan commented on src/controllers/changePassword.ts:
```diff
@@ -0,0 +1,90 @@
+  secretOrKey: secret,
+};
+
+passport.use(
+  new JwtStrategy(jwtOptions, async (jwtPayload: jwtPayloadType, done) => {
+    try {
+      const userId = await Database.User.findOne({
+        where: {
+          id: jwtPayload.id,
+        },
+      });
+      logger.info(userId);
+      if (!userId) return done(null, false);
+      return done(null, userId);
+    } catch (err) {
+      logger.info(err);
+      return done(err, false);
+    }
+  })
+);
```
Please create strategy to different file

fab-ryan commented on src/controllers/changePassword.ts:
```diff
@@ -0,0 +1,90 @@
+      try {
+        if (err) {
+          return res.status(400).json({ status: 400, error: 'login first' });
+        }
+        if (user) {
+          const checkErrors = changePasswordValidation.validate(req.body, {
+            abortEarly: false,
+          });
+          if (checkErrors.error) {
+            const messageError = checkErrors.error.details
+              // eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-explicit-any
+              .map((error: any) => error.message)
+              .join(', ');
+            return res.status(400).json({ status: 400, error: messageError });
+          }
+          logger.info('----------------------', checkErrors);
+          const { oldPassword, newPassword } = req.body;
+
+          const checkOldPassword = await bcrypt.compare(
+            oldPassword,
+            user.password
+          );
```
use validation as middlware not like this

fab-ryan reviewed (changes requested):
first rebase to develop branch

fab-ryan commented on src/controllers/loginController.ts:
```diff
@@ -0,0 +1,66 @@
+  }
+}
+
+passport.use(
+  new LocalStrategy({ usernameField: 'email' }, checkUserCredentials)
+);
+
+const validatingUser = async (req: Request, res: Response) => {
+  passport.authenticate(
+    'local',
+    { session: false },
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    async (err: any, user: any) => {
+      // test if uer not found
+
+      if (!user) return sendResponse<null>(res, 400, null, 'User not found');
+
+      // generate token
+
+      const { secret } = config();
+      const token = jwt.sign({ id: user.id }, secret, { expiresIn: '2h' });
+
```
I told you to separate this with current codes

Ruthuwamahoro replied:
The only way I did this is to remove that strategy and put it in app.ts.Could you explain further?

fab-ryan commented on src/controllers/changePasswordController.ts:
```diff
@@ -0,0 +1,75 @@
+  email: string;
+  password: string;
+}
+
+passport.use('jwt', defineJwtStrategy);
+
+/// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+const updatePassword = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  passport.authenticate(
+    'jwt',
+    { session: false },
+    async (err: Error | null, user: User) => {
+      try {
+        // if (err) {
+        //   return sendResponse<null>(res, 400, null, 'user not found');
+        // }
+        if (user) {
+          const { oldPassword, newPassword } = req.body;
```
same here remove those stategy put it in app.ts and first create it's files you use it in global

Ruthuwamahoro replied:
I have created generalStrategy.ts in helper folder and I made changes.

fab-ryan reviewed (commented):
please optimize your codes

fab-ryan commented on src/controllers/changePasswordController.ts:
```diff
@@ -0,0 +1,63 @@
+  res: Response,
+  next: NextFunction
+) => {
+  return passport.authenticate(
```
this are handled in middlwares

RWEMAREMY reviewed (approved):
good job

fab-ryan commented on src/middleware/validatePassword.ts:
```diff
@@ -0,0 +1,56 @@
+import { NextFunction, Request, Response } from 'express';
+import Joi from 'joi';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { logger, sendResponse } from '../utils';
+
+// eslint-disable-next-line no-shadow
+enum requestType {
+  body = 'body',
+  params = 'params',
+  queries = 'queries',
+}
+
+interface errorInterface {
+  field: string;
+  message: string;
+}
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
+export const validationMiddleware =
+  (schema: Joi.ObjectSchema, type: requestType = requestType.body) =>
+  (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const { error } = schema.validate(req.body, { abortEarly: false });
+
+      if (error) {
+        const errors: errorInterface[] = [];
+        error.details.forEach(err => {
+          const field = err.path[0] as string;
+          if (field === 'password') {
```
this file is not necessary we already have validationn middleware use that

fab-ryan commented on src/routers/userRoute.ts:
```diff
@@ -16,5 +19,10 @@ router.post(
   validationMiddleware(loginUserSchema),
   validatingUser
 );
+router.post(
```
use validation middlware only pass schemas

fab-ryan commented on test/updatePassword.test.ts:
```diff
@@ -0,0 +1,100 @@
+    const user = await chai.request(app).post('/api/users/login').send({
+      email: 'test1@example.com',
+      password: userPassword.oldPassword,
+    });
+
+    const Token: string = user.body.data;
+
+    chai
+      .request(app)
+      .post('/api/users/password')
+      .set('Authorization', `Bearer ${Token}`)
+      .send(userPassword)
+      .end((err, res) => {
+        expect(res).to.have.status(200);
+        expect(res.body).to.have.property('message');
+        expect(res.body.message).to.equal('change Password Successfully');
+      });
+  });
+  it('when old password is not the same as the one stored in database', async () => {
```
no test that trigger 500

fab-ryan reviewed (changes requested):
Work on the requests

fab-ryan commented on src/middleware/authentication.ts:
```diff
@@ -11,14 +11,15 @@ export const isAuthenticated = (
         id: user.id,
         name: user.name,
         email: user.email,
+        password: user.password,
```
we don't need password please remove it  you have id of user please remove password

fab-ryan commented on src/routers/userRoute.ts:
```diff
@@ -16,5 +19,11 @@ router.post(
   validationMiddleware(loginUserSchema),
   validatingUser
```
please cchange the name of this controller to loginController

fab-ryan reviewed (changes requested):
Please correct those mistakes

Ruthuwamahoro committed:
feat(update-profile):user should be able to update profile

Ensuring that the user attempting to change the password is a valid registered user with a validated account
Requires the user to provide their current password for authentication purposes, along with the new password they want to set and confirmation to avoid errors

[finished #187300151]

Ruthuwamahoro committed:
-amend
